Addad data simulator service

ofek29 opened the pull request:
Added data simulator service with its CI workflow for setup and testing.
Also added Kafka broker setup in the CI.

ofek29 committed:
Addad data simulatur service

ofek29 committed:
Fixed kafka auto create topic

avifenesh commented:
Why you have the ci test file twice?
Dont call it CI, call it something like "data_sim_test", you'll have billion CI files.

ofek29 committed:
Fixed tests

ofek29 commented:
Fixed some issues with the test configuration.
Improved test flow: Use different topics and consumers, Folder Structure, Admin API use for managing topics.
Changed the CI to start on pr and push to main.
Please review

avifenesh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,81 @@
+name: test_data_simulator
```
I meant to the name of the file. For having separate CI running for each service, you can't call one of them CI, give it a name that reflects the usage.

ofek29 replied:
Changed

avifenesh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,81 @@
+      - main
+  pull_request:
+    branches:
+      - main
```
Add path as well. You don't want the whole CI to run on each change, just on relevant changes.
For example, if you change the README you want a CI that checks the formatting, but you don't need to run the DS tests etc.

avifenesh replied:
https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#example-including-paths

ofek29 replied:
I added paths for push and pr

avifenesh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,81 @@
+          - 9092:9092
+          - 29092:29092
+          - 29093:29093
+        options: >-
```
try to use env
https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#example-of-env

ofek29 replied:
Changed to env

avifenesh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,81 @@
+        image: confluentinc/cp-enterprise-control-center:7.8.0
+        ports:
+          - 9021:9021
+        options: >-
```
same

ofek29 replied:
Changed

avifenesh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,81 @@
+
+      - name: Wait for Kafka to be ready
+        run: |
+          timeout 120 sh -c '
```
Use timeout for the step

ofek29 replied:
I added health check for the broker insted

avifenesh commented on data-simulator/kafka_producer.py:
```diff
@@ -0,0 +1,30 @@
+print("Producer instance created.")
+
+# Delivery callback for produced messages
+def delivery_callback(err, msg):
```
You can't print on each message, it has a very high price to call stdout so many times. Use an efficient logger, and add logs base on level set. In prod, you log just warning, in dev you log info level, if you need to debug you use trace or debug level.

avifenesh replied:
Do so over whole code please.

ofek29 replied:
I setup logger as you instructed, using log level for prod and debug.
Also added app running types to support this.

avifenesh commented on kafka/docker-compose.yml:
```diff
@@ -0,0 +1,47 @@
+services:
```
If the service intends to use the compose you should test the real environment in the CI, why do use the automatic?

avifenesh commented on data-simulator/tests/test_sim_engine.py:
```diff
@@ -0,0 +1,60 @@
+
+    # Check if the topic exists, otherwise create it
+    existing_topics = admin_client.list_topics(timeout=10).topics
+    if topic_name not in existing_topics:
```
I'm not sure if I understand the reason to use if. Should it create the topic, or should the topic exist already?

ofek29 replied:
Needed when topic doesn't exist. 
But i removed it and enabled kafka auto create topic

avifenesh commented on data-simulator/tests/test_sim_engine.py:
```diff
@@ -0,0 +1,60 @@
+        new_topic = NewTopic(topic_name, num_partitions=1, replication_factor=1)
+        admin_client.create_topics([new_topic])
+        print(f"Topic '{topic_name}' created.")
+        time.sleep(2)  # Wait for the topic to become available
```
that's will be flaky. Use a deterministic way to check if the topic is available.

avifenesh commented on data-simulator/tests/test_sim_engine.py:
```diff
@@ -0,0 +1,60 @@
+
+
+def test_data_simulator_integration(setup_kafka, kafka_consumer):
+    num_cars = 3
```
use global

ofek29 replied:
I changed all related variables to global

avifenesh commented on data-simulator/tests/test_sim_engine.py:
```diff
@@ -0,0 +1,60 @@
+    'auto.offset.reset': 'earliest'
+}
+
+@pytest.fixture
```
fixture can be set to run per module, per function etc., set the fixture base on what you need.

avifenesh commented on data-simulator/tests/test_sim_engine.py:
```diff
@@ -0,0 +1,60 @@
+import json
```
Add at least one more test, if test can be two threads, and you can run simultaneously its better.

avifenesh commented on data-simulator/tests/test_producer_consumer.py:
```diff
@@ -0,0 +1,64 @@
+import json
```
What are you testing? This is a producer service. Test the producer. When testing the producer, you need to consume, so you'll have a consumer for the tests but no reason to test it especially.

avifenesh commented on .github/workflows/ci.yml:
Please use mypy over your code, and a formatter and linter of choice. There's black, pep8 and a few more.
Chose the one that fits your preferences, but you need to have styling consistency.
After choosing, fix code style and types, and add mypy and the formatter checking to the CI as well.

ofek29 replied:
Added mypy and forrmatter in my ide and in the CI as you suggested

avifenesh commented on data-simulator/simulator.py:
```diff
@@ -0,0 +1,25 @@
+        for i in range(4):
+            self.generate_data()
+            time.sleep(2)
+        # while True:
```
Remove the commented code

ofek29 replied:
Fixed with the implementation of app running levels

avifenesh commented on data-simulator/simulator.py:
```diff
@@ -0,0 +1,25 @@
+            print(f"Sent telemetry data for car {car.id} with success: {success}")
+
+    def run(self):
+        for i in range(4):
```
I don't understand the logic here

ofek29 replied:
This is the number of laps, when on dev mode
Change to global to be more clear.

avifenesh commented on data-simulator/models.py:
```diff
@@ -0,0 +1,14 @@
+import random
```
I don't think we need models here. You can choose, it's not important, but I think the class can be part of the simulator file.

ofek29 replied:
I change it to car file instead, I thought this looks more clear.
I'll check the convection and move it if necessary.

avifenesh commented on data-simulator/main.py:
```diff
@@ -0,0 +1,11 @@
+from simulator import DataSimulator
+
+simulator = DataSimulator(5)
```
Why 5?

ofek29 replied:
I change it to const variable, so i can change it easily if needed
This represents the number of car racing on the track.

avifenesh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,81 @@
+
+      - name: Install dependencies
+        run: |
+          cd data-simulator
```
Use working directory parameter.

ofek29 replied:
Fixed

avifenesh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,81 @@
+      - name: Install dependencies
+        run: |
+          cd data-simulator
+          python -m pip install --upgrade pip
```
Why update?

avifenesh replied:
anyway, use caching for speeding up and doing all together 
https://github.com/actions/setup-python#caching-packages-dependencies

ofek29 replied:
> Why update?

removed according to instruction

ofek29 replied:
Still need to understand why caching is failing..

avifenesh commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,81 @@
+
+      - name: Run tests
+        run: |
+          cd data-simulator
```
Same, working directory.

ofek29 replied:
Fixed

avifenesh reviewed (commented):
Left comments inside. Don't get overwhelmed, we are learning.

ofek29 committed:
Added formatter and static typing for python

ofek29 committed:
Added kafka broker health check

ofek29 committed:
Changed Broker setup environment variables

ofek29 committed:
Added logger with diffrent log levels

ofek29 committed:
Changed client fixture to run once

ofek29 committed:
Added app run mods, Change to global variable

ofek29 committed:
Fix path

ofek29 committed:
Fix path

ofek29 committed:
Fix path

ofek29 committed:
Fix path

ofek29 committed:
Fixed fixture

ofek29 committed:
Fixed test setup

ofek29 committed:
Fixed test setup

ofek29 committed:
Fixed test setup

ofek29 committed:
Improve kafka service setup

ofek29 committed:
Changed check kafka workflow

ofek29 committed:
Specify module name

ofek29 committed:
Improve logger setup, create only if not exist

ofek29 committed:
Change variable to const

ofek29 committed:
New e2e test for the service

ofek29 committed:
Fix utils as module

ofek29 committed:
Delete unnecessary and changed files

ofek29 committed:
Styling
